Merge the opensea-erc1155 code into opensea-creatures

robmyers opened the pull request:
Add contracts, migrations, tests, scripts and support code from opensea-erc1155 to opensea-creatures to unify the repos.

Notes:

* This commit adds the opensea-erc1155 contracts, migrations, tests, scripts and configuration to opensea-creatures.
* It merges the README.md files and flatten.sh files.
* It changes truffle.js to include the local test support from opensea-erc1155/truffle.js .
* It updates package.json and package-lock.json to include the libraries required by the new tests.
* The MyXXX contracts from opensea-erc1155 have been renamed CreatureAccessoryXXX.
* ERC1155 scripts are in scripts/creature-accessories.
* The new lib/ directory contains shared code for the test/*.json files.
* There was a clash between the mint() method of opensea-creatures Factory.sol and opensea-ERC1155 IFactory.sol . See CreatureAccessoryFactory.sol for details.

dfinzer commented on contracts/CreatureAccessoryFactory.sol:
```diff
@@ -0,0 +1,173 @@
+//FIXME: ERC-721 Factory and ERC-1155 IFactory have clashing definitions of
+// mint(), so we do *not* implement Factory as otherwise we'd need the two-param
+// version of mint(). 
+//import "./Factory.sol";
```
Gotcha. Just a note that when this is alll stable, we should try to deprecate the old version of `Factory` to avoid confusion. We should also use [EIP 165](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md) instead of `supportsFactoryInterface`

robmyers replied:
This may indicate deeper problems in making the ERC1155 version of the presale tutorial. I'm digging into it.

robmyers replied:
I've resolved this for now.

robmyers commented on contracts/CreatureAccessory.sol:
```diff
@@ -0,0 +1,20 @@
+pragma solidity ^0.5.11;
+
+import "./ERC1155Tradable.sol";
```
Done in e91d23c .

robmyers commented:
Regarding the conflicts:
* TradeableERC721Token.sol is now ERC721Tradable.sol .
* The migrations have been modified to use async and support CreatureAccessories.
* package.json and the resuling package-lock.json have had support libraries added.

robmyers commented:
OK here are the last changes. Please note that I have modified the scripts to use the correct ABI but I cannot test them locally as they fail to connect to the web API silently:
1. The LootBoxes are now actual 1155 tokens.
2. The Factory code supports pre-minted and mint-on-demand tokens. Accesories are the former, to support the tutorial, and LootBoxes are the latter, to demonstrate that it works.
3. The lootbox code wouldn't fit into the space left once we subclass ERC1155Tradable, so I made it a library and linked that to CreatureAccessoryLootBox. This and separating the LootBox from the Factory code makes things look slightly more complex, although it's pretty much the same amount of code.
4. Due to needing to pre-mint tokens, transfer ownership of contracts, call approveForAll(), etc. I moved the contract configuration JavaScript to /lib/. It's called by the tests and migrations.

dfinzer commented on lib/valuesCommon.js:
```diff
@@ -0,0 +1,57 @@
+
+const NUM_ACCESSORIES = 6;
+const MINT_INITIAL_SUPPLY = 1000;
+const INITIAL_SUPPLY = toBNHex('0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF');
```
Could use constant from above file.

robmyers replied:
Sorry, I don't quite follow. :-) In which context? MINT_INITIAL_SUPPLY is used in the migration and test code to pre-mint the tokens, INITIAL_SUPPLY reflects the value in CreatureAccessoryFactory that is the limit on lazily minted tokens, it's currently UINT256_MAX for both.

dfinzer replied:
Oh sorry nvm, thought the const `0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF` was defined elsewhere.

dfinzer commented on migrations/2_deploy_contracts.js:
```diff
@@ -1,23 +1,88 @@
-  // })
-};
\ No newline at end of file
+  if (DEPLOY_CREATURES) {
```
Thanks for adding env variables!

robmyers replied:
You are welcome! :-)

dfinzer commented on contracts/CreatureAccessoryFactory.sol:
```diff
@@ -0,0 +1,188 @@
+  ) public {
+    proxyRegistryAddress = _proxyRegistryAddress;
+    nftAddress = _nftAddress;
+    lootBoxAddress = _lootBoxAddress;
```
If it's not too much effort, could we fire `TransferSingle` events for each option, similar to [what we do for the CreatureFactory](https://github.com/ProjectOpenSea/opensea-creatures/blob/master/contracts/CreatureFactory.sol#L38)?

robmyers replied:
mint() will emit either safeTransferFrom() (for pre-minted tokens) or _mint() (for lazily minted tokens).
Both emit TransferSingle, although it is emitted from their respective ERC1155 contract rather than from CreatureAccessoryFactory.
If you would like this added to TransferSingle so there's a central place to watch for it that makes sense, although it is in theory redundant.
Let me know. :-)

dfinzer replied:
I was actuallly talking about [in the constructor](https://github.com/ProjectOpenSea/opensea-creatures/blob/master/contracts/CreatureFactory.sol#L38) so that our backend picks up the _existence_ of the options, but it's a bit tricky so let's hold off.

robmyers commented on contracts/LootBoxRandomness.sol:
```diff
@@ -0,0 +1,325 @@
+pragma solidity ^0.5.11;
+
+// Eek!
```
I was importing OpenZeppelin's SafeMath and it was breaking, so I had to import multi-token-standard's SafeMath, which seemed like exposing a project's internals. The comment is not constructive though, I have removed it.

robmyers commented on test/CreatureAccessoryFactory.js:
```diff
@@ -0,0 +1,235 @@
+    // which is the correct behavior (no reentrancy) for the wrong reason
+    // (the attacker is not the owner or proxy).
+
+    xit('Minting from factory should disallow re-entrancy attack',
```
The x disables the test. This is simpler to toggle than commenting the code out.

I can't find "xit" in the mocha docs but it's discussed here.

https://stackoverflow.com/a/32724129/989917

dfinzer replied:
Gotcha, thanks!

robmyers commented on README.md:
```diff
@@ -69,4 +69,103 @@ If you're running a modified version of `sell.js` and not getting expected behav
+# About OpenSea Creature Accessories
+
+This is a sample ERC-1155 contract for the purposes of demonstrating integration with the [OpenSea](https://opensea.io) marketplace for crypto collectibles. We also include:
+- A script for minting items.
```
Correct. I have removed this reference.

The unboxing/lootbox open script in scripts/creature-accessories/advanced is called "mint.js" though.

robmyers commented on README.md:
```diff
@@ -69,4 +69,103 @@ If you're running a modified version of `sell.js` and not getting expected behav
+export OWNER_ADDRESS="<my_address>"
+export NFT_CONTRACT_ADDRESS="<deployed_contract_address>"
+export NETWORK="rinkeby"
+node scripts/mint.js
```
Agreed. I have removed this.

robmyers commented on contracts/CreatureAccessoryFactory.sol:
```diff
@@ -0,0 +1,188 @@
+    uint256 _optionId
+  ) public view returns (uint256) {
+    if ( _optionId < NUM_ITEM_OPTIONS) {
+      if ((!_isOwnerOrProxy(_owner)) && _owner != lootBoxAddress) {
```
Yes, left over from working though a bug. Thank you for catching that! I have removed them.

robmyers commented on contracts/ERC1155Tradable.sol:
```diff
@@ -0,0 +1,223 @@
+    *       creating NFTs by setting the id's high bits with the method
+    *       described in ERC1155 or to use ids representing values other than
+    *       successive small integers. If you wish to create ids as successive
+    *       small integers you can either subclass this class to cound onchain
```
Eek! Fixed, thank you.

robmyers commented on contracts/IFactory.sol:
```diff
@@ -0,0 +1,66 @@
+  // Get things to work on OpenSea with mock methods below
+  ///////
+
+  //function safeTransferFrom(address _from, address _to, uint256 _optionId, uint256 _amount, bytes calldata _data) external;
```
The Factory now creates both CreatureAccesory and CreatureAccessoryLootBox tokens as ERC1155Tradable tokens.
These tokens can be listed and transferred without the CreatureAccessoryFactory having to support safeTransferFrom.
I therefore did not implement it on CreatureAccessoryFactory and commented it out here.

robmyers commented on contracts/ERC1155Tradable.sol:
```diff
@@ -0,0 +1,223 @@
+    creators[_id] = msg.sender;
+
+    if (bytes(_uri).length > 0) {
+      emit URI(_uri, _id);
```
Correct. I think this method is trying to work nicely with ERC1155's use of the event log as a source of information. But the contract does support on-chain URIs. So I've added a store for custom URIs, a method for the token's creator to update them, and code for uri() to properly use them.

robmyers commented on contracts/CreatureAccessoryFactory.sol:
```diff
@@ -0,0 +1,188 @@
+    address _owner,
+    uint256 _optionId
+  ) public view returns (uint256) {
+    if ( _optionId < NUM_ITEM_OPTIONS) {
```
Fixed! Thank you.
